lib/puppet/util/package/version/rpm.rb

```diff
@@ -46,12 +39,35 @@ def eql?(other)
     def <=>(other)
       raise ArgumentError, _("Cannot compare, as %{other} is not a Rpm Version") % { other: other } unless other.is_a?(self.class)
 
-      cmp = @epoch <=> other.epoch
-      if cmp == 0
-        cmp = rpm_compareEVR(rpm_parse_evr(self.to_s), rpm_parse_evr(other.to_s))
-      end
-      cmp
+      rpm_compareEVR(self.to_s, other.to_s)
     end
 
+    private
+
+    # overwrite rpm_compareEVR to treat no epoch as zero epoch
+    # in order to compare version correctly
+    #
+    # returns 1 if a is newer than b,
+    #         0 if they are identical
+    #        -1 if a is older than b
+    def rpm_compareEVR(a, b)
+      a_hash = rpm_parse_evr(a)
+      b_hash = rpm_parse_evr(b)
+
+      a_hash[:epoch] ||= '0'
+      b_hash[:epoch] ||= '0'
```

Comment:
With this change, "1.2" will be translated to "0:1.2" and will no longer match "<any>:1.2".
For example, if version "2:1.2" is available, "ensure=1.2" will be no-op, while "ensure='<=1.2'" will fail, as it will search for packages "<=0:1.2"
This change and https://github.com/puppetlabs/puppet/blob/86dbe5bad71d96a1995615d17ceaa8b1f01828cb/lib/puppet/provider/package/yum.rb#L209 are not symmetric

Reply:
I think this is expected as we don't want to make assumptions if no epoch is provided by the user, we default to using the rpm comparison algorithm.